text/0000-fn-shorthand.md

```diff
@@ -0,0 +1,477 @@
+sense as units are also good candidates:
+
+```rust
+fn compile_error(msg: &str) -> TokenStream = quote! { compile_error!(#msg); };
+```
+
+Additionally, using `try { .. }` as the immediate expression of a function
+becomes quite ergonomic, for example:
+
+```rust
+#[post("/github-webhook", data = "<event>")]
+pub fn github_webhook(event: Event) -> DashResult<()> = try {
+    let conn = &*DB_POOL.get()?;
+
+    match event.payload {
+        Payload::Issues(issue_event) => { /* logic... */ },
+        Payload::PullRequest(pr_event) => { /* logic... */ },
+        Payload::IssueComment(comment_event) => { /* logic... */ },
+    }
+
+    // Notice the lack of `Ok(())` because of `try { .. }`.
+};
+```
+
+This becomes a cheap way to starve off some of the needs for a `try fn` construct.
+
+### In the standard library
+
+As this is a language changes that is quite widely and generally applicable,
+enumerating all the places where it would apply would be unfeasible.
+However, one instance of where it applies frequently is in `src/core/iter/mod.rs`.
+For example:
+
+```rust
+impl<I: Iterator> Iterator for Peekable<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<I::Item> = match self.peeked.take() {
+        Some(v) => v,
+        None => self.iter.next(),
+    };
+
+    fn count(mut self) -> usize = match self.peeked.take() {
+        Some(None) => 0,
+        Some(Some(_)) => 1 + self.iter.count(),
+        None => self.iter.count(),
+    };
+
+    ...
+}
+```
+
+## Encouraging small function definitions
+
+[SRP]: https://en.wikipedia.org/wiki/Single_responsibility_principle
+[SoC]: https://en.wikipedia.org/wiki/Separation_of_concerns
+
+As a result of reduced rightward drift, and removing the need to always
+have 2 extra lines for the braces, as seen in the definition of `double`,
+we can make it more ergonomic to write small function definitions. This in
+turn can in many cases improve readability as it becomes easier to define self
+contained units that respect the *[single responsibility principle (SRP)][SRP]*,
+*[separation of concerns (SoC)][SoC]* or *modularity* in general.
+
+## Embracing the expression oriented nature
+
+[expr_oriented]: https://doc.rust-lang.org/book/first-edition/glossary.html#expression-oriented-language
+
+As the book tells us, [Rust is an expression-oriented language][expr_oriented].
+Therefore it is somewhat peculiar that one of the most important parts
+of the language, the function definition body, is not an expression as
+in other expression oriented languages (see the [prior-art]).
+In this RFC, we make the function body into an expression.
+
+## A more syntactically unified feeling
+
+Finally, a rather minor motivation, which still bears mentioning, is
+that by permitting `fn foo(...) -> R = expr;` as a syntax, we bring `const`
+and `static` items closer to `fn` items syntactically which gives a smoother
+experience overall. In other words, we can now write:
```

Comment:
I disagree that this gives a smoother experience. I think this is confusing, as static, consts and fns work differently. Having equality makes it seem like functions should be eagerly evaluated.

Reply:
This is not really the place to debate the proposal itself but...
I think that's quite a stretch; in none of the languages noted in the prior art does such confusion occur as there's clearly an argument that needs to be provided to the function before anything happens, so how could it possibly be eagerly evaluated?

Reply:
Some of the languages I would argue are not analogous to Rust (e.g. Haskell / any lazy language), but regardless, a function with no arguments could plausibly be eagerly evaluated. In many of the languages, `=` is the default way to specify a function, which I think makes it less confusing.

Reply:
+ A function with no arguments is really a function that is `() -> ReturnType` so it still takes an argument.
+ In a function with no arguments you are also more likely to write `fn main() { ... }` because the purpose of such a function is primarily its side effects.
+ Also, most functional programming languages, e.g. ML, OCaml, Scala, and Idris are all strict but no such confusion occurs.
+ Scala, Kotlin, and C# all support both `{ ... }` and `=` (or `=>` in C#) and there doesn't seem to arise such confusion from this. Notably the Scala users that use Rust seem to like this idea from what I can tell.